# Getting data from reuters

In [1]:
from reuterspy import Reuters
reuters = Reuters()
import pandas as pd
pd.set_option('display.max_rows', None)


### Method for reshaping reuters data

In [2]:
def reshape_reuters_statements(df):
    return df.pivot(index="metric", columns="year", values="value").reset_index()

In [3]:
def get_metric_latest_bs(df, metric_name):
    return int(float(df[2020][df['metric'] == metric_name].iloc[0]))

# Selecting data

In [4]:
ticker_list = ['ABB.N']

## Get cash flow

In [5]:
df_cf = reuters.get_cash_flow(ticker_list)
df_cf = reshape_reuters_statements(df_cf)

KeyError: 'metric'

In [6]:
df_cf

year                                  metric          2015          2016  \
0                       Capital Expenditures   -876.000000   -632.000000   
1                         Cash Interest Paid    221.000000    213.000000   
2                            Cash Taxes Paid   1043.000000    814.000000   
3             Cash from Financing Activities  -3380.000000  -3355.000000   
4             Cash from Investing Activities   -974.000000  -1305.000000   
5             Cash from Operating Activities   3818.000000   3843.000000   
6                 Changes in Working Capital    658.000000    528.000000   
7                             Deferred Taxes   -219.000000   -145.000000   
8                     Depreciation/Depletion   1160.000000    870.000000   
9                  Financing Cash Flow Items   -613.000000  -1773.000000   
10                  Foreign Exchange Effects   -342.000000   -104.000000   
11        Issuance (Retirement) of Debt, Net    -30.000000   -475.000000   
12       Issuance (Retirement) of Stock, Net  -1380.000000  -1107.000000   
13                        Net Change in Cash   -878.000000   -921.000000   
14                  Net Income/Starting Line   2055.000000   2034.000000   
15                            Non-Cash Items    164.000000    556.000000   
16    Other Investing Cash Flow Items, Total    -98.000000   -673.000000   
17                 Total Cash Dividends Paid  -1357.000000      0.000000   

year          2017          2018          2019          2020  
0      -752.000000   -772.000000   -762.000000   -694.000000  
1       205.000000    243.000000    284.000000    189.000000  
2       894.000000   1026.000000   1005.000000    905.000000  
3     -1735.000000   -789.000000  -1383.000000  -8175.000000  
4     -1450.000000  -3085.000000   -815.000000   6760.000000  
5      3799.000000   2924.000000   2325.000000   1693.000000  
6       639.000000     50.000000   -372.000000    402.000000  
7      -199.000000   -142.000000    -83.000000   -280.000000  
8       836.000000    916.000000    961.000000    915.000000  
9      -136.000000   -169.000000   -132.000000   -100.000000  
10      268.000000   -131.000000    -28.000000     79.000000  
11      124.000000   1305.000000    414.000000  -3703.000000  
12      -88.000000   -208.000000     10.000000  -2636.000000  
13      882.000000  -1081.000000     99.000000    357.000000  
14     2365.000000   2298.000000   1528.000000   5205.000000  
15      158.000000   -198.000000    291.000000  -4549.000000  
16     -698.000000  -2313.000000    -53.000000   7454.000000  
17    -1635.000000  -1717.000000  -1675.000000  -1736.000000

## Get balancesheet data

In [6]:
df_bs = reuters.get_balance_sheet(ticker_list)
df_bs = reshape_reuters_statements(df_bs)

KeyError: 'metric'

In [8]:
df_bs

year                                    metric          2015          2016  \
0                             Accounts Payable   4342.000000   4446.000000   
1             Accounts Receivable - Trade, Net  10324.000000  10037.000000   
2                             Accrued Expenses           NaN   2641.000000   
3              Accumulated Depreciation, Total  -6840.000000  -6926.000000   
4                   Additional Paid-In Capital   1175.000000     24.000000   
5                    Capital Lease Obligations           NaN           NaN   
6                                         Cash   1837.000000   1704.000000   
7                           Cash & Equivalents   2717.000000   1940.000000   
8              Cash and Short Term Investments   6466.000000   5853.000000   
9                          Common Stock, Total    269.000000    192.000000   
10    Current Port. of  LT Debt/Capital Leases   1176.000000    868.000000   
11                         Deferred Income Tax    965.000000    918.000000   
12                               Goodwill, Net   9671.000000   9501.000000   
13                            Intangibles, Net   2337.000000   1996.000000   
14                              Long Term Debt   5985.000000   5800.000000   
15                       Long Term Investments    263.000000    282.000000   
16                           Minority Interest    507.000000    502.000000   
17                 Note Receivable - Long Term           NaN    112.000000   
18               Notes Payable/Short Term Debt    278.000000    135.000000   
19                 Other Current Assets, Total   1251.000000    980.000000   
20            Other Current liabilities, Total  10048.000000   7059.000000   
21                         Other Equity, Total  -4865.000000  -5194.000000   
22                    Other Liabilities, Total   3574.000000   3438.000000   
23               Other Long Term Assets, Total   1049.000000   1516.000000   
24                            Prepaid Expenses    225.000000    176.000000   
25     Property/Plant/Equipment, Total - Gross  12116.000000  11669.000000   
26       Property/Plant/Equipment, Total - Net   5276.000000   4743.000000   
27     Retained Earnings (Accumulated Deficit)  20476.000000  19925.000000   
28                      Short Term Investments   1912.000000   2209.000000   
29    Tangible Book Value per Share, Common Eq      1.128390      0.887450   
30                                Total Assets  41356.000000  39202.000000   
31             Total Common Shares Outstanding   2191.625140   2138.706840   
32                        Total Current Assets  22760.000000  21052.000000   
33                   Total Current Liabilities  15844.000000  15149.000000   
34                                  Total Debt   7439.000000   6803.000000   
35                                Total Equity  14481.000000  13395.000000   
36                             Total Inventory   4757.000000   4347.000000   
37                           Total Liabilities  26875.000000  25807.000000   
38    Total Liabilities & Shareholders' Equity  41356.000000  39202.000000   
39                        Total Long Term Debt   5985.000000   5800.000000   
40                      Total Receivables, Net  10061.000000   9696.000000   
41                     Treasury Stock - Common  -2581.000000  -1559.000000   
42                      Unrealized Gain (Loss)      7.000000      7.000000   

year          2017          2018          2019          2020  
0      3736.000000   4424.000000   4353.000000   4571.000000  
1      6492.000000   6833.000000   6764.000000   7045.000000  
2      2433.000000   2735.000000   2942.000000   3061.000000  
3     -5547.000000  -5528.000000  -5716.000000  -6364.000000  
4        29.000000     56.000000     73.000000     83.000000  
5              NaN           NaN    154.000000    160.000000  
6      1963.000000   1983.000000   2111.000000   1765.000000  
7      2563.000000   1462.000000   1433.000000   1513.000000  


## Income statement

In [9]:
df_is = reuters.get_income_statement(ticker_list)
df_is = reshape_reuters_statements(df_is)
df_is

year                                    metric          2015          2016  \
0                       Cost of Revenue, Total  25347.000000  17384.000000   
1             DPS - Common Stock Primary Issue      0.740000      0.750580   
2         Diluted EPS Excluding ExtraOrd Items      0.865470      0.510680   
3                           Diluted Net Income   1933.000000   1899.000000   
4                       Diluted Normalized EPS      1.153190      0.650030   
5              Diluted Weighted Average Shares   2230.000000   2154.000000   
6                          Dilution Adjustment      0.000000      0.000000   
7                                 Gross Profit  10134.000000   7545.000000   
8        Income Available to Com Excl ExtraOrd   1930.000000   1100.000000   
9        Income Available to Com Incl ExtraOrd   1933.000000   1899.000000   
10    Interest Exp.(Inc.),Net-Operating, Total     -2.000000    -38.000000   
11      Interest Inc.(Exp.),Net-Non-Op., Total   -209.000000   -130.000000   
12                           Minority Interest   -122.000000   -135.000000   
13                                  Net Income   1933.000000   1899.000000   
14                      Net Income After Taxes   2052.000000   1235.000000   
15              Net Income Before Extra. Items   1930.000000   1100.000000   
16                     Net Income Before Taxes   2840.000000   1761.000000   
17                            Operating Income   3049.000000   1929.000000   
18             Other Operating Expenses, Total           NaN           NaN   
19                                  Other, Net           NaN    -38.000000   
20                  Provision for Income Taxes    788.000000    526.000000   
21                      Research & Development   1406.000000    956.000000   
22                                     Revenue  35481.000000  24929.000000   
23      Selling/General/Admin. Expenses, Total   5587.000000   4534.000000   
24                   Total Extraordinary Items      3.000000    799.000000   
25                     Total Operating Expense  32432.000000  23000.000000   
26                               Total Revenue  35481.000000  24929.000000   
27                    Unusual Expense (Income)     94.000000    164.000000   

year          2017          2018          2019          2020  
0     17184.000000  19059.000000  19018.000000  18123.000000  
1         0.762580      0.805990      0.786310      0.825210  
2         0.634540      0.677890      0.488520      0.138740  
3      2213.000000   2173.000000   1439.000000   5146.000000  
4         0.706690      0.787000      0.633570      0.376870  
5      2148.000000   2139.000000   2135.000000   2119.000000  
6         0.000000      0.000000      0.000000      0.000000  
7      8012.000000   8603.000000   8960.000000   8011.000000  
8      1363.000000   1450.000000   1043.000000    294.000000  
9      2213.000000   2173.000000   1439.000000   5146.000000  
10       69.000000    -52.000000    -33.000000      8.000000  
11     -161.000000   -190.000000   -148.000000   -189.000000  
12     -152.000000   -125.000000    -47.000000    -51.000000  
13     2213.000000   2173.000000   1439.000000   5146.000000  
14     1515.000000   1575.000000   1090.000000    345.000000  
15     1363.000000   1450.000000   1043.000000    294.000000  
16     2102.000000   2119.000000   1862.000000    841.000000  
17     2230.000000   2226.000000   1938.000000   1431.000000  
18             NaN           NaN    392.000000   -116.000000  
19       33.000000     83.000000     72.000000   -401.000000  
20      587.000000    544.000000    772.000000    496.000000  
21     1013.000000   1141.000000   1198.000000   1111.000000  
22    25196.000000  27662.000000  27978.000000  26134.000000  
23     4735.000000   5232.000000   5399.000000   4797.000000  
24      850.000000    723.000000    396.000000   4852.000000  
25    22966.000000  25436.000000  26040.000000  24703.000000  
26    25196.000000  27662.000000  27

### EBIT = Revenue − COGS − Operating Expenses

* [Kaavan lähde](https://www.investopedia.com/terms/e/ebit.asp)
==> Revenue - Cost of Revenue, Total - Total Operating Expense

In [10]:
df_is_orig = reuters.get_income_statement(ticker_list)
df_is_orig

ticker   financialReport  year                                    metric  \
0    ABB.N  income_statement  2020                                   Revenue   
1    ABB.N  income_statement  2019                                   Revenue   
2    ABB.N  income_statement  2018                                   Revenue   
3    ABB.N  income_statement  2017                                   Revenue   
4    ABB.N  income_statement  2016                                   Revenue   
5    ABB.N  income_statement  2015                                   Revenue   
6    ABB.N  income_statement  2020                             Total Revenue   
7    ABB.N  income_statement  2019                             Total Revenue   
8    ABB.N  income_statement  2018                             Total Revenue   
9    ABB.N  income_statement  2017                             Total Revenue   
10   ABB.N  income_statement  2016                             Total Revenue   
11   ABB.N  income_statement  2015                             Total Revenue   
12   ABB.N  income_statement  2020                    Cost of Revenue, Total   
13   ABB.N  income_statement  2019                    Cost of Revenue, Total   
14   ABB.N  income_statement  2018                    Cost of Revenue, Total   
15   ABB.N  income_statement  2017                    Cost of Revenue, Total   
16   ABB.N  income_statement  2016                    Cost of Revenue, Total   
17   ABB.N  income_statement  2015                    Cost of Revenue, Total   
18   ABB.N  income_statement  2020                              Gross Profit   
19   ABB.N  income_statement  2019                              Gross Profit   
20   ABB.N  income_statement  2018                              Gross Profit   
21   ABB.N  income_statement  2017                              Gross Profit   
22   ABB.N  income_statement  2016                              Gross Profit   
23   ABB.N  income_statement  2015                              Gross Profit   
24   ABB.N  income_statement  2020    Selling/General/Admin. Expenses, Total   
25   ABB.N  income_statement  2019    Selling/General/Admin. Expenses, Total   
26   ABB.N  income_statement  2018    Selling/General/Admin. Expenses, Total   
27   ABB.N  income_statement  2017    Selling/General/Admin. Expenses, Total   
28   ABB.N  income_statement  2016    Selling/General/Admin. Expenses, Total   
29   ABB.N  income_statement  2015    Selling/General/Admin. Expenses, Total   
30   ABB.N  income_statement  2020                    Research & Development   
31   ABB.N  income_statement  2019                    Research & Development   
32   ABB.N  income_statement  2018                    Research & Development   
33   ABB.N  income_statement  2017                    Research & Development   
34   ABB.N  income_statement  2016                    Research & Development   
35   ABB.N  income_statement  2015                    Research & Development   
36   ABB.N  income_statement  2020  Interest Exp.(Inc.),Net-Operating, Total   
37   ABB.N  income_statement  2019  Interest Exp.(Inc.),Net-Operating, Total   
38   ABB.N  income_statement  2018  Interest Exp.(Inc.),Net-Operating, Total   
39   ABB.N  income_statement  2017  Interest Exp.(Inc.),Net-Operating, Total   
40   ABB.N  income_statement  2016  Interest Exp.(Inc.),Net-Operating, Total   
41   ABB.N  income_statement  2015  Interest Exp.(Inc.),Net-Operating, Total   
42   ABB.N  income_statement  2020                  Unusual Expense (Income)   
43   ABB.N  income_statement  2019                  Unusual Expense (Income)   
44   ABB.N  income_statement  2018                  Unusual Expense (Income)   
45   ABB.N  income_statement  2017                  Unusual Expense (Income)   
46   ABB.N  income_statement  2016                  Unusual Expense (Income)   
47   ABB.N  income_statement  2015                  Unusual Expense (Income)   
48   ABB.N  income_statement  2020           Other Operating Expenses, Total   
49  

## EBIT KAAVA:

Revenue - Cost of Revenue, Total - Selling/General/Admin. Expenses, Total + Interest Inc.(Exp.),Net-Non-Op., Total

## key metrics

In [12]:
df_key_metrics = reuters.get_key_metrics(ticker_list)
df_key_metrics

ticker                                             metric       value  \
0    ABB.N                          Price closing or last bid       34.95   
1    ABB.N                                       52 Week High       38.03   
2    ABB.N                                        52 Week Low       24.07   
3    ABB.N                                       Pricing date  2021-09-21   
4    ABB.N                      10 Day Average Trading Volume        1.15   
5    ABB.N                              Market Capitalization   71,558.88   
6    ABB.N                                  52 Week High Date  2021-08-24   
7    ABB.N                                   52 Week Low Date  2020-10-30   
8    ABB.N                     3 Month Average Trading Volume       25.74   
9    ABB.N                                               Beta        1.10   
10   ABB.N                                 1 Day Price Change        0.78   
11   ABB.N                       13 Week Price Return (Daily)        3.28   
12   ABB.N                       26 Week Price Return (Daily)       12.63   
13   ABB.N                         5 Day Price Return (Daily)       -5.05   
14   ABB.N                       52 Week Price Return (Daily)       38.58   
15   ABB.N                  Year To Date Price Return (Daily)       25.00   
16   ABB.N                 Month to Date Price Return (Daily)       -5.52   
17   ABB.N                  Price Relative to S&P500 (4 Week)       -5.11   
18   ABB.N                 Price Relative to S&P500 (13 Week)        0.72   
19   ABB.N                 Price Relative to S&P500 (26 Week)        1.16   
20   ABB.N                 Price Relative to S&P500 (52 Week)        4.43   
21   ABB.N                     Price Relative to S&P500 (YTD)        7.83   
22   ABB.N                     EPS excl. Extra Items (Annual)        0.14   
23   ABB.N                        EPS excl. Extra Items (TTM)        0.44   
24   ABB.N                            EPS Normalized (Annual)        0.38   
25   ABB.N                         Revenue per Share (Annual)       12.33   
26   ABB.N                            Revenue per Share (TTM)       13.65   
27   ABB.N                      Book Value (Per Share Annual)        7.72   
28   ABB.N                   Book Value (Per Share Quarterly)        6.95   
29   ABB.N             Tangible Book Value (Per Share Annual)        1.36   
30   ABB.N          Tangible Book Value (Per Share Quarterly)        0.63   
31   ABB.N                  Cash Per Share (Per Share Annual)        2.80   
32   ABB.N               Cash Per Share (Per Share Quarterly)        1.99   
33   ABB.N                       Cash Flow (Per Share Annual)        0.59   
34   ABB.N                          Cash Flow (Per Share TTM)        0.91   
35   ABB.N                        Dividend (Per Share Annual)        0.83   
36   ABB.N                          Dividends (Per Share TTM)        0.83   
37   ABB.N                              EBITD (Per Share TTM)        2.17   
38   ABB.N               EPS Basic excl. Extra Items (Annual)        0.14   
39   ABB.N                  EPS Basic excl. Extra Items (TTM)        0.45   
40   ABB.N                     EPS incl. Extra Items (Annual)        2.43   
41   ABB.N                        EPS incl. Extra Items (TTM)        2.72   
42   ABB.N                     Free Cash Flow (Per Share TTM)        0.48   
43   ABB.N                            Dividend (Per Share 5Y)        0.79   
44   ABB.N                     P/E excl. Extra Items (Annual)      251.91   
45   ABB.N                        P/E excl. Extra Items (TTM)       78.85   
46   ABB.N                            P/E Normalized (Annual)       92.74   
47   ABB.N                            Price to sales (Annual)        2.74   
48   ABB.N                               Price to sales (TTM)        2.55   
49   ABB.N                    Price to Tangible Book (Annual)       25.96   
50   ABB.N                 Price to Tangible Book (Quarterly)       56.93   


# Reshape data

In [13]:
## Already done here, but here's how you could do it.
# df_bs = reshape_reuters_statements(df_bs)
# df_is = reshape_reuters_statements(df_is)
# ==> Doesn't have yearly data: df_key_metrics = reshape_reuters_statements(df_key_metrics)

In [14]:
df_bs

year                                    metric          2015          2016  \
0                             Accounts Payable   4342.000000   4446.000000   
1             Accounts Receivable - Trade, Net  10324.000000  10037.000000   
2                             Accrued Expenses           NaN   2641.000000   
3              Accumulated Depreciation, Total  -6840.000000  -6926.000000   
4                   Additional Paid-In Capital   1175.000000     24.000000   
5                    Capital Lease Obligations           NaN           NaN   
6                                         Cash   1837.000000   1704.000000   
7                           Cash & Equivalents   2717.000000   1940.000000   
8              Cash and Short Term Investments   6466.000000   5853.000000   
9                          Common Stock, Total    269.000000    192.000000   
10    Current Port. of  LT Debt/Capital Leases   1176.000000    868.000000   
11                         Deferred Income Tax    965.000000    918.000000   
12                               Goodwill, Net   9671.000000   9501.000000   
13                            Intangibles, Net   2337.000000   1996.000000   
14                              Long Term Debt   5985.000000   5800.000000   
15                       Long Term Investments    263.000000    282.000000   
16                           Minority Interest    507.000000    502.000000   
17                 Note Receivable - Long Term           NaN    112.000000   
18               Notes Payable/Short Term Debt    278.000000    135.000000   
19                 Other Current Assets, Total   1251.000000    980.000000   
20            Other Current liabilities, Total  10048.000000   7059.000000   
21                         Other Equity, Total  -4865.000000  -5194.000000   
22                    Other Liabilities, Total   3574.000000   3438.000000   
23               Other Long Term Assets, Total   1049.000000   1516.000000   
24                            Prepaid Expenses    225.000000    176.000000   
25     Property/Plant/Equipment, Total - Gross  12116.000000  11669.000000   
26       Property/Plant/Equipment, Total - Net   5276.000000   4743.000000   
27     Retained Earnings (Accumulated Deficit)  20476.000000  19925.000000   
28                      Short Term Investments   1912.000000   2209.000000   
29    Tangible Book Value per Share, Common Eq      1.128390      0.887450   
30                                Total Assets  41356.000000  39202.000000   
31             Total Common Shares Outstanding   2191.625140   2138.706840   
32                        Total Current Assets  22760.000000  21052.000000   
33                   Total Current Liabilities  15844.000000  15149.000000   
34                                  Total Debt   7439.000000   6803.000000   
35                                Total Equity  14481.000000  13395.000000   
36                             Total Inventory   4757.000000   4347.000000   
37                           Total Liabilities  26875.000000  25807.000000   
38    Total Liabilities & Shareholders' Equity  41356.000000  39202.000000   
39                        Total Long Term Debt   5985.000000   5800.000000   
40                      Total Receivables, Net  10061.000000   9696.000000   
41                     Treasury Stock - Common  -2581.000000  -1559.000000   
42                      Unrealized Gain (Loss)      7.000000      7.000000   

year          2017          2018          2019          2020  
0      3736.000000   4424.000000   4353.000000   4571.000000  
1      6492.000000   6833.000000   6764.000000   7045.000000  
2      2433.000000   2735.000000   2942.000000   3061.000000  
3     -5547.000000  -5528.000000  -5716.000000  -6364.000000  
4        29.000000     56.000000     73.000000     83.000000  
5              NaN           NaN    154.000000    160.000000  
6      1963.000000   1983.000000   2111.000000   1765.000000  
7      2563.000000   1462.000000   1433.000000   1513.000000  


In [15]:
# df_bs['2020'][df_bs['metric'] == 'Total Current Assets']#[df_bs['metric'] == 'Total Current Assets'].values[0]

### Get total current assets

In [16]:
total_current_assets = int(float(df_bs[2020][df_bs['metric'] == 'Total Current Assets'].iloc[0]))
total_current_liabilities = get_metric_latest_bs(df_bs, 'Total Current Liabilities')
print(int(total_current_assets))
print(type(total_current_liabilities))
# Working capital = total_current_assets - total_current_liabilities
working_capital = total_current_assets - total_current_liabilities
print(working_capital)

total_assets = get_metric_latest_bs(df_bs, 'Total Assets')
print(total_assets)

A = working_capital/(float(total_assets))
print(A)

19226
<class 'int'>
3810
41088
0.09272780373831775


In [17]:
# B
retained_earnings = get_metric_latest_bs(df_bs, 'Retained Earnings (Accumulated Deficit)')
retained_earnings
B = retained_earnings/(float(total_assets))
B

0.5584598909657321

In [18]:
# C = Earnings before interest / total assets
ebit = get_metric_latest_bs(df_is, 'Revenue') - get_metric_latest_bs(df_is, 'Total Operating Expense')
C = ebit / float(total_assets)
print(C)

0.03482768691588785


## D = Market Value of Equity / Total Liabilities

In [19]:
import yfinance as yf
pd.set_option('display.float_format', str)
ticker = yf.Ticker("AMZN")



In [20]:
from yahoofinancials import YahooFinancials

ticker = 'ABB'
yahoo_financials = YahooFinancials(ticker)

# Hakee hinnan
yahoo_financials.get_prev_close_price()

# prev close 22.9.2021
price = yahoo_financials.get_prev_close_price()

ticker_yfinance = yf.Ticker(ticker)
shares = ticker_yfinance.info['sharesOutstanding']

# Kaikki muut arvot ilmoitettu miljoonissa
market_value_of_equity = (price*shares)/(1000000)

liabilities = get_metric_latest_bs(df_bs,'Total Liabilities')
print(liabilities)

D = market_value_of_equity/float(liabilities)
print("D: ", D)

E = get_metric_latest_bs(df_is, 'Revenue') / float(get_metric_latest_bs(df_bs, 'Total Assets'))
print("E: ", E)

25403
D:  2.7598984371924575
E:  0.6360494548286605


In [21]:
Z = (1.2*A)+(1.4*B)+(3.3*C)+(0.6*D)+(1.0*E)
print("Z: ", Z)

Z:  3.3000370958045706


# Z-Score:

* AMZN: 6.150276669527528

In [146]:
df_key_metrics



ticker                                             metric         value  \
0    AMZN.O                          Price closing or last bid      3,462.52   
1    AMZN.O                                       52 Week High      3,773.08   
2    AMZN.O                                        52 Week Low      2,871.00   
3    AMZN.O                                       Pricing date    2021-09-17   
4    AMZN.O                      10 Day Average Trading Volume          2.79   
5    AMZN.O                              Market Capitalization  1,753,560.00   
6    AMZN.O                                  52 Week High Date    2021-07-13   
7    AMZN.O                                   52 Week Low Date    2020-09-21   
8    AMZN.O                     3 Month Average Trading Volume         68.35   
9    AMZN.O                                               Beta          1.15   
10   AMZN.O                                 1 Day Price Change         -0.74   
11   AMZN.O                       13 Week Price Return (Daily)         -0.70   
12   AMZN.O                       26 Week Price Return (Daily)         12.60   
13   AMZN.O                         5 Day Price Return (Daily)         -0.19   
14   AMZN.O                       52 Week Price Return (Daily)         15.08   
15   AMZN.O                  Year To Date Price Return (Daily)          6.31   
16   AMZN.O                 Month to Date Price Return (Daily)         -0.24   
17   AMZN.O                  Price Relative to S&P500 (4 Week)          8.42   
18   AMZN.O                 Price Relative to S&P500 (13 Week)         -6.67   
19   AMZN.O                 Price Relative to S&P500 (26 Week)         -0.60   
20   AMZN.O                 Price Relative to S&P500 (52 Week)        -12.85   
21   AMZN.O                     Price Relative to S&P500 (YTD)         -9.92   
22   AMZN.O                     EPS excl. Extra Items (Annual)         41.83   
23   AMZN.O                        EPS excl. Extra Items (TTM)         57.38   
24   AMZN.O                            EPS Normalized (Annual)         61.70   
25   AMZN.O                         Revenue per Share (Annual)        756.99   
26   AMZN.O                            Revenue per Share (TTM)        864.13   
27   AMZN.O                      Book Value (Per Share Annual)        185.69   
28   AMZN.O                   Book Value (Per Share Quarterly)        226.88   
29   AMZN.O             Tangible Book Value (Per Share Annual)        145.94   
30   AMZN.O          Tangible Book Value (Per Share Quarterly)        196.55   
31   AMZN.O                  Cash Per Share (Per Share Annual)        167.79   
32   AMZN.O               Cash Per Share (Per Share Quarterly)        177.66   
33   AMZN.O                       Cash Flow (Per Share Annual)         74.56   
34   AMZN.O                          Cash Flow (Per Share TTM)        115.02   
35   AMZN.O                        Dividend (Per Share Annual)            --   
36   AMZN.O                          Dividends (Per Share TTM)          0.00   
37   AMZN.O                              EBITD (Per Share TTM)        137.08   
38   AMZN.O               EPS Basic excl. Extra Items (Annual)         42.66   
39   AMZN.O                  EPS Basic excl. Extra Items (TTM)         58.57   
40   AMZN.O                     EPS incl. Extra Items (Annual)         41.83   
41   AMZN.O                        EPS incl. Extra Items (TTM)         57.38   
42   AMZN.O                     Free Cash Flow (Per Share TTM)         13.78   
43   AMZN.O                            Dividend (Per Share 5Y)            --   
44   AMZN.O                     P/E excl. Extra Items (Annual)         82.78   
45   AMZN.O                        P/E excl. Extra Items (TTM)         60.35   
46   AMZN.O                            P/E Normalized (Annual)         56.11   
47   AMZN.O                            Price to sales (Annual)          4.54   
48   AMZN.O                               Price to sales (TTM)          3.96   
49   

In [117]:
df_is

year                                  metric           2015           2016  \
0                     Cost of Revenue, Total   71651.000000   88265.000000   
1           DPS - Common Stock Primary Issue            NaN       0.000000   
2                  Depreciation/Amortization     228.000000     287.000000   
3       Diluted EPS Excluding ExtraOrd Items       1.249480       4.898760   
4                         Diluted Net Income     596.000000    2371.000000   
5                     Diluted Normalized EPS       1.249480       4.898760   
6            Diluted Weighted Average Shares     477.000000     484.000000   
7                        Dilution Adjustment       0.000000            NaN   
8                       Equity In Affiliates     -22.000000     -96.000000   
9                               Gross Profit   35355.000000   47722.000000   
10     Income Available to Com Excl ExtraOrd     596.000000    2371.000000   
11     Income Available to Com Incl ExtraOrd     596.000000    2371.000000   
12    Interest Inc.(Exp.),Net-Non-Op., Total    -680.000000    -371.000000   
13                                Net Income     596.000000    2371.000000   
14                    Net Income After Taxes     618.000000    2467.000000   
15            Net Income Before Extra. Items     596.000000    2371.000000   
16                   Net Income Before Taxes    1568.000000    3892.000000   
17                          Operating Income    2233.000000    4186.000000   
18           Other Operating Expenses, Total     -57.000000    -120.000000   
19                                Other, Net      15.000000      77.000000   
20                Provision for Income Taxes     950.000000    1425.000000   
21                    Research & Development   12540.000000   16085.000000   
22                                   Revenue  107006.000000  135987.000000   
23    Selling/General/Admin. Expenses, Total   20411.000000   27284.000000   
24                 Total Extraordinary Items            NaN            NaN   
25                   Total Operating Expense  104773.000000  131801.000000   
26                             Total Revenue  107006.000000  135987.000000   

year           2017           2018           2019           2020  
0     111934.000000  139156.000000  165536.000000  233307.000000  
1               NaN            NaN            NaN            NaN  
2        366.000000     475.000000     565.000000     509.000000  
3          4.551720      19.832000      22.992060      41.825490  
4       3033.000000   10073.000000   11588.000000   21331.000000  
5          4.551720      19.832000      22.992060      61.704400  
6        493.000000     500.000000     504.000000     510.000000  
7               NaN            NaN            NaN            NaN  
8         -4.000000       9.000000     -14.000000      16.000000  
9      65932.000000   93731.000000  114986.000000  152757.000000  
10      2244.000000    9916.000000   11588.000000   21331.000000  
11      3033.000000   10073.000000   11588.000000   21331.000000  
12      -406.000000    -977.000000    -557.000000    -224.000000  
13      3033.000000   10073.000000   11588.000000   21331.000000  
14      2248.000000    9907.000000   11602.000000   21315.000000  
15      2244.000000    9916.000000   11588.000000   21331.000000  
16      3806.000000   11261.000000   13976.000000   24178.000000  
17      4106.000000   12421.000000   14541.000000   22899.000000  
18      -152.000000    -179.000000    -364.000000    -584.000000  
19       106.000000    -183.000000      -8.000000    1503.000000  
20      1558.000000    1354.000000    2374.000000    2863.000000  
21     22620.000000   28837.000000   35931.000000   42740.000000  
22    177866.000000  232887.000000  280522.000000  386064.000000  
23     38992.000000   52177.000000   64313.000000   87193.000000  
24       789.000000     157.000000       0.000000       0.000000  
25    173760.000000  220466.000000  265981.000000  363165.000000  
26    

In [115]:
df_key_metrics

ticker                                             metric         value  \
0    AMZN.O                          Price closing or last bid      3,462.52   
1    AMZN.O                                       52 Week High      3,773.08   
2    AMZN.O                                        52 Week Low      2,871.00   
3    AMZN.O                                       Pricing date    2021-09-20   
4    AMZN.O                      10 Day Average Trading Volume          2.79   
5    AMZN.O                              Market Capitalization  1,753,560.00   
6    AMZN.O                                  52 Week High Date    2021-07-13   
7    AMZN.O                                   52 Week Low Date    2020-09-21   
8    AMZN.O                     3 Month Average Trading Volume         68.35   
9    AMZN.O                                               Beta          1.15   
10   AMZN.O                                 1 Day Price Change         -0.74   
11   AMZN.O                       13 Week Price Return (Daily)         -0.70   
12   AMZN.O                       26 Week Price Return (Daily)         12.60   
13   AMZN.O                         5 Day Price Return (Daily)         -0.19   
14   AMZN.O                       52 Week Price Return (Daily)         15.08   
15   AMZN.O                  Year To Date Price Return (Daily)          6.31   
16   AMZN.O                 Month to Date Price Return (Daily)         -0.24   
17   AMZN.O                  Price Relative to S&P500 (4 Week)          8.42   
18   AMZN.O                 Price Relative to S&P500 (13 Week)         -6.67   
19   AMZN.O                 Price Relative to S&P500 (26 Week)         -0.60   
20   AMZN.O                 Price Relative to S&P500 (52 Week)        -12.85   
21   AMZN.O                     Price Relative to S&P500 (YTD)         -9.92   
22   AMZN.O                     EPS excl. Extra Items (Annual)         41.83   
23   AMZN.O                        EPS excl. Extra Items (TTM)         57.38   
24   AMZN.O                            EPS Normalized (Annual)         61.70   
25   AMZN.O                         Revenue per Share (Annual)        756.99   
26   AMZN.O                            Revenue per Share (TTM)        864.13   
27   AMZN.O                      Book Value (Per Share Annual)        185.69   
28   AMZN.O                   Book Value (Per Share Quarterly)        226.88   
29   AMZN.O             Tangible Book Value (Per Share Annual)        145.94   
30   AMZN.O          Tangible Book Value (Per Share Quarterly)        196.55   
31   AMZN.O                  Cash Per Share (Per Share Annual)        167.79   
32   AMZN.O               Cash Per Share (Per Share Quarterly)        177.66   
33   AMZN.O                       Cash Flow (Per Share Annual)         74.56   
34   AMZN.O                          Cash Flow (Per Share TTM)        115.02   
35   AMZN.O                        Dividend (Per Share Annual)            --   
36   AMZN.O                          Dividends (Per Share TTM)          0.00   
37   AMZN.O                              EBITD (Per Share TTM)        137.08   
38   AMZN.O               EPS Basic excl. Extra Items (Annual)         42.66   
39   AMZN.O                  EPS Basic excl. Extra Items (TTM)         58.57   
40   AMZN.O                     EPS incl. Extra Items (Annual)         41.83   
41   AMZN.O                        EPS incl. Extra Items (TTM)         57.38   
42   AMZN.O                     Free Cash Flow (Per Share TTM)         13.78   
43   AMZN.O                            Dividend (Per Share 5Y)            --   
44   AMZN.O                     P/E excl. Extra Items (Annual)         82.78   
45   AMZN.O                        P/E excl. Extra Items (TTM)         60.35   
46   AMZN.O                            P/E Normalized (Annual)         56.11   
47   AMZN.O                            Price to sales (Annual)          4.54   
48   AMZN.O                               Price to sales (TTM)          3.96   
49   

In [78]:
value = get_metric_latest_bs(df_bs, 'Total Current Assets')
print(value)

29    132733.000000
Name: 2020, dtype: object


In [116]:
df_cf

year                                  metric          2015          2016  \
0                       Capital Expenditures  -5387.000000  -7804.000000   
1                         Cash Interest Paid    325.000000    290.000000   
2                            Cash Taxes Paid    273.000000    412.000000   
3             Cash from Financing Activities  -3882.000000  -3716.000000   
4             Cash from Investing Activities  -6450.000000  -9516.000000   
5             Cash from Operating Activities  12039.000000  17203.000000   
6                 Changes in Working Capital   2557.000000   3847.000000   
7                             Deferred Taxes     81.000000   -246.000000   
8                     Depreciation/Depletion   6281.000000   8116.000000   
9                   Foreign Exchange Effects   -374.000000   -212.000000   
10        Issuance (Retirement) of Debt, Net  -3882.000000  -3716.000000   
11                        Net Change in Cash   1333.000000   3759.000000   
12                  Net Income/Starting Line    596.000000   2371.000000   
13                            Non-Cash Items   2524.000000   3115.000000   
14    Other Investing Cash Flow Items, Total  -1063.000000  -1712.000000   

year           2017           2018           2019           2020  
0     -11955.000000  -13427.000000  -16861.000000  -40140.000000  
1        328.000000     854.000000     875.000000     916.000000  
2        957.000000    1184.000000    -881.000000    1713.000000  
3       9928.000000   -7686.000000  -10066.000000   -1104.000000  
4     -27084.000000  -12369.000000  -24281.000000  -59611.000000  
5      18365.000000   30723.000000   38514.000000   66064.000000  
6       -242.000000   -1043.000000   -2438.000000   13481.000000  
7        -29.000000     441.000000     796.000000    -554.000000  
8      11478.000000   15341.000000   21789.000000   25251.000000  
9        713.000000    -351.000000      70.000000     618.000000  
10      9928.000000   -7686.000000  -10066.000000   -1104.000000  
11      1922.000000   10317.000000    4237.000000    5967.000000  
12      3033.000000   10073.000000   11588.000000   21331.000000  
13      4125.000000    5911.000000    6779.000000    6555.000000  
14    -15129.000000    1058.000000   -7420.000000  -19471.000000

In [136]:
import yfinance as yf
pd.set_option('display.float_format', str)
msft = yf.Ticker("AMZN")
msft.financials


2020-12-31     2019-12-31  \
Research Development                    42740000000.0  35931000000.0   
Effect Of Accounting Charges                     None           None   
Income Before Tax                       24194000000.0  13962000000.0   
Minority Interest                                None           None   
Net Income                              21331000000.0  11588000000.0   
Selling General Administrative          87193000000.0  64313000000.0   
Gross Profit                           152757000000.0 114986000000.0   
Ebit                                    22899000000.0  14404000000.0   
Operating Income                        22899000000.0  14404000000.0   
Other Operating Expenses                  -75000000.0    338000000.0   
Interest Expense                        -1647000000.0  -1600000000.0   
Extraordinary Items                              None           None   
Non Recurring                                    None           None   
Other Items                                      None           None   
Income Tax Expense                       2863000000.0   2374000000.0   
Total Revenue                          386064000000.0 280522000000.0   
Total Operating Expenses               363165000000.0 266118000000.0   
Cost Of Revenue                        233307000000.0 165536000000.0   
Total Other Income Expense Net           1295000000.0   -442000000.0   
Discontinued Operations                          None           None   
Net Income From Continuing Ops          21331000000.0  11588000000.0   
Net Income Applicable To Common Shares  21331000000.0  11588000000.0   

                                           2018-12-31     2017-12-31  
Research Development                    28837000000.0  22620000000.0  
Effect Of Accounting Charges                     None           None  
Income Before Tax                       11270000000.0   3802000000.0  
Minority Interest                                None           None  
Net Income                              10073000000.0   3033000000.0  
Selling General Administrative          52177000000.0  38992000000.0  
Gross Profit                            93731000000.0  65932000000.0  
Ebit                                    12421000000.0   4106000000.0  
Operating Income                        12421000000.0   4106000000.0  
Other Operating Expenses                  296000000.0    214000000.0  
Interest Expense                        -1417000000.0   -848000000.0  
Extraordinary Items                              None           None  
Non Recurring                                    None           None  
Other Items                                      None           None  
Income Tax Expense                       1197000000.0    769000000.0  
Total Revenue                          232887000000.0 177866000000.0  
Total Operating Expenses               220466000000.0 173760000000.0  
Cost Of Revenue                        139156000000.0 111934000000.0  
Total Other Income Expense Net          -1151000000.0   -304000000.0  
Discontinued Operations                          None           None  
Net Income From Continuing Ops          10073000000.0   3033000000.0  
Net Income Applicable To Common Shares  10073000000.0   3033000000.0

In [137]:
msft.cashflow

2020-12-31     2019-12-31  \
Investments                               -22242000000.0  -9131000000.0   
Change To Liabilities                      18745000000.0   9904000000.0   
Total Cashflows From Investing Activities -59611000000.0 -24281000000.0   
Net Borrowings                             -1104000000.0 -10066000000.0   
Total Cash From Financing Activities       -1104000000.0 -10066000000.0   
Change To Operating Activities              5754000000.0  -1383000000.0   
Net Income                                 21331000000.0  11588000000.0   
Change In Cash                              5967000000.0   4237000000.0   
Effect Of Exchange Rate                      618000000.0     70000000.0   
Total Cash From Operating Activities       66064000000.0  38514000000.0   
Depreciation                               25251000000.0  21789000000.0   
Change To Inventory                        -2849000000.0  -3278000000.0   
Change To Account Receivables              -8169000000.0  -7681000000.0   
Change To Netincome                         6001000000.0   7575000000.0   
Capital Expenditures                      -40140000000.0 -16861000000.0   
Other Cashflows From Investing Activities            nan            nan   

                                              2018-12-31     2017-12-31  
Investments                                 1140000000.0  -3054000000.0  
Change To Liabilities                       4414000000.0   7838000000.0  
Total Cashflows From Investing Activities -12369000000.0 -27084000000.0  
Net Borrowings                             -7686000000.0   9928000000.0  
Total Cash From Financing Activities       -7686000000.0   9928000000.0  
Change To Operating Activities               472000000.0    283000000.0  
Net Income                                 10073000000.0   3033000000.0  
Change In Cash                             10317000000.0   1922000000.0  
Effect Of Exchange Rate                     -351000000.0    713000000.0  
Total Cash From Operating Activities       30723000000.0  18365000000.0  
Depreciation                               15341000000.0  11478000000.0  
Change To Inventory                        -1314000000.0  -3583000000.0  
Change To Account Receivables              -4615000000.0  -4780000000.0  
Change To Netincome                         6352000000.0   4096000000.0  
Capital Expenditures                      -13427000000.0 -11955000000.0  
Other Cashflows From Investing Activities   2104000000.0   1897000000.0

In [138]:
msft.earnings


Revenue     Earnings
Year                           
2017  177866000000   3033000000
2018  232887000000  10073000000
2019  280522000000  11588000000
2020  386064000000  21331000000

# YahooFinancials lisäosalla

* [Linkki](https://github.com/JECSand/yahoofinancials)

In [29]:
from yahoofinancials import YahooFinancials

ticker = 'CAT'
yahoo_financials = YahooFinancials(ticker)

balance_sheet_data_qt = yahoo_financials.get_financial_stmts('annual', 'balance')
income_statement_data_qt = yahoo_financials.get_financial_stmts('annual', 'income')
all_statement_data_qt =  yahoo_financials.get_financial_stmts('annual', ['income', 'cash', 'balance'])
apple_earnings_data = yahoo_financials.get_stock_earnings_data()
apple_net_income = yahoo_financials.get_net_income()
historical_stock_prices = yahoo_financials.get_historical_price_data('2008-09-15', '2018-09-15', 'weekly')

yahoo_financials.get_prev_close_price()

balance_sheet_data_qt


{'balanceSheetHistory': {'CAT': [{'2020-12-31': {'intangibleAssets': 1308000000,
     'totalLiab': 62946000000,
     'totalStockholderEquity': 15331000000,
     'minorityInterest': 47000000,
     'otherCurrentLiab': 3561000000,
     'totalAssets': 78324000000,
     'commonStock': 6230000000,
     'otherCurrentAssets': 133000000,
     'retainedEarnings': 35167000000,
     'otherLiab': 10773000000,
     'goodWill': 6394000000,
     'treasuryStock': -26066000000,
     'otherAssets': 4180000000,
     'cash': 8822000000,
     'totalCurrentLiabilities': 25717000000,
     'deferredLongTermAssetCharges': 1358000000,
     'shortLongTermDebt': 1420000000,
     'otherStockholderEquity': -888000000,
     'propertyPlantEquipment': 8912000000,
     'totalCurrentAssets': 39464000000,
     'longTermInvestments': 1687000000,
     'netTangibleAssets': 7629000000,
     'netReceivables': 7178000000,
     'longTermDebt': 9646000000,
     'inventory': 11402000000,
     'accountsPayable': 6128000000}},
   {'

In [ ]:
print("income statement")
print(income_statement_data_qt)

income statement
{'incomeStatementHistory': {'ABB': [{'2020-12-31': {'researchDevelopment': 1113000000, 'effectOfAccountingCharges': None, 'incomeBeforeTax': 841000000, 'minorityInterest': 314000000, 'netIncome': 5146000000, 'sellingGeneralAdministrative': 5144000000, 'grossProfit': 8011000000, 'ebit': 1881000000, 'operatingIncome': 1881000000, 'otherOperatingExpenses': -127000000, 'interestExpense': -448000000, 'extraordinaryItems': None, 'nonRecurring': None, 'otherItems': None, 'incomeTaxExpense': 496000000, 'totalRevenue': 26134000000, 'totalOperatingExpenses': 24253000000, 'costOfRevenue': 18123000000, 'totalOtherIncomeExpenseNet': -1040000000, 'discontinuedOperations': 4860000000, 'netIncomeFromContinuingOps': 345000000, 'netIncomeApplicableToCommonShares': 5146000000}}, {'2019-12-31': {'researchDevelopment': 1197000000, 'effectOfAccountingCharges': None, 'incomeBeforeTax': 1862000000, 'minorityInterest': 454000000, 'netIncome': 1439000000, 'sellingGeneralAdministrative': 5135000

In [ ]:
balance_sheet_data_qt

{'balanceSheetHistory': {'ABB': [{'2020-12-31': {'intangibleAssets': 2077000000,
     'capitalSurplus': 83000000,
     'totalLiab': 25089000000,
     'totalStockholderEquity': 15685000000,
     'minorityInterest': 314000000,
     'otherCurrentLiab': 6313000000,
     'totalAssets': 41088000000,
     'commonStock': 188000000,
     'otherCurrentAssets': 1357000000,
     'retainedEarnings': 22946000000,
     'otherLiab': 4114000000,
     'goodWill': 10850000000,
     'treasuryStock': -7532000000,
     'otherAssets': 1930000000,
     'cash': 3278000000,
     'totalCurrentLiabilities': 15416000000,
     'deferredLongTermAssetCharges': 843000000,
     'shortLongTermDebt': 1115000000,
     'otherStockholderEquity': -4002000000,
     'propertyPlantEquipment': 5143000000,
     'totalCurrentAssets': 19226000000,
     'longTermInvestments': 1862000000,
     'netTangibleAssets': 2758000000,
     'shortTermInvestments': 2108000000,
     'netReceivables': 7805000000,
     'longTermDebt': 4668000000,


In [154]:
from yahoofinancials import YahooFinancials

ticker = 'AMZN'
yahoo_financials = YahooFinancials(ticker)


import yfinance as yf

ticker_yfinance = yf.Ticker(ticker)

# Hakee osakkeiden lukumäärän
shares_outstanding = ticker_yfinance.info['sharesOutstanding']

# Hakee hinnan
prev_close = yahoo_financials.get_prev_close_price()

market_value_of_equity = shares_outstanding*prev_close
print("mkt eqt: ", market_value_of_equity)

#total_liabilities = 

mkt eqt:  1693351294080.96


In [153]:
ticker_yfinance.info['sharesOutstanding']

506440992

In [135]:
income_statement_data_qt

{'incomeStatementHistory': {'AMZN': [{'2020-12-31': {'researchDevelopment': 42740000000,
     'effectOfAccountingCharges': None,
     'incomeBeforeTax': 24194000000,
     'minorityInterest': None,
     'netIncome': 21331000000,
     'sellingGeneralAdministrative': 87193000000,
     'grossProfit': 152757000000,
     'ebit': 22899000000,
     'operatingIncome': 22899000000,
     'otherOperatingExpenses': -75000000,
     'interestExpense': -1647000000,
     'extraordinaryItems': None,
     'nonRecurring': None,
     'otherItems': None,
     'incomeTaxExpense': 2863000000,
     'totalRevenue': 386064000000,
     'totalOperatingExpenses': 363165000000,
     'costOfRevenue': 233307000000,
     'totalOtherIncomeExpenseNet': 1295000000,
     'discontinuedOperations': None,
     'netIncomeFromContinuingOps': 21331000000,
     'netIncomeApplicableToCommonShares': 21331000000}},
   {'2019-12-31': {'researchDevelopment': 35931000000,
     'effectOfAccountingCharges': None,
     'incomeBeforeTax': 

In [128]:
balance_sheet_data_qt

{'balanceSheetHistoryQuarterly': {'AAPL': [{'2021-06-26': {'totalLiab': 265560000000,
     'totalStockholderEquity': 64280000000,
     'otherCurrentLiab': 51306000000,
     'totalAssets': 329840000000,
     'commonStock': 54989000000,
     'otherCurrentAssets': 13641000000,
     'retainedEarnings': 9233000000,
     'otherLiab': 38354000000,
     'treasuryStock': 58000000,
     'otherAssets': 44854000000,
     'cash': 34050000000,
     'totalCurrentLiabilities': 107754000000,
     'shortLongTermDebt': 8039000000,
     'otherStockholderEquity': 58000000,
     'propertyPlantEquipment': 38615000000,
     'totalCurrentAssets': 114423000000,
     'longTermInvestments': 131948000000,
     'netTangibleAssets': 64280000000,
     'shortTermInvestments': 27646000000,
     'netReceivables': 33908000000,
     'longTermDebt': 105752000000,
     'inventory': 5178000000,
     'accountsPayable': 40409000000}},
   {'2021-03-27': {'totalLiab': 267980000000,
     'totalStockholderEquity': 69178000000,
   

In [130]:
print(all_statement_data_qt)

{'incomeStatementHistoryQuarterly': {'AAPL': [{'2021-06-26': {'researchDevelopment': 5717000000, 'effectOfAccountingCharges': None, 'incomeBeforeTax': 24369000000, 'minorityInterest': None, 'netIncome': 21744000000, 'sellingGeneralAdministrative': 5412000000, 'grossProfit': 35255000000, 'ebit': 24126000000, 'operatingIncome': 24126000000, 'otherOperatingExpenses': None, 'interestExpense': -665000000, 'extraordinaryItems': None, 'nonRecurring': None, 'otherItems': None, 'incomeTaxExpense': 2625000000, 'totalRevenue': 81434000000, 'totalOperatingExpenses': 57308000000, 'costOfRevenue': 46179000000, 'totalOtherIncomeExpenseNet': 243000000, 'discontinuedOperations': None, 'netIncomeFromContinuingOps': 21744000000, 'netIncomeApplicableToCommonShares': 21744000000}}, {'2021-03-27': {'researchDevelopment': 5262000000, 'effectOfAccountingCharges': None, 'incomeBeforeTax': 28011000000, 'minorityInterest': None, 'netIncome': 23630000000, 'sellingGeneralAdministrative': 5314000000, 'grossProfit':

# Calculating Ponsse's beta using different indices

In [39]:
#!pip install pandas-datareader

In [61]:
import numpy as np
import pandas as pd
from pandas_datareader import data

tickers = ["^OMXHCAPGI", 'PON1V.HE']


In [62]:
#import yfinance as yf
tickers = ["^OMXHCAPGI", "PON1V.HE"]
#omxh = yf.Ticker(tickers)
#hist = omxh.history(period="5y", interval='1d')

from pandas_datareader import data

beta_data = data.DataReader(tickers, 'yahoo', start='2016-11-25', end='2021-11-25')


In [63]:
beta_data = beta_data['Adj Close']

In [72]:
beta_data.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 1258 entries, 2016-11-25 to 2021-11-25
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   ^OMXHCAPGI  1248 non-null   float64
 1   PON1V.HE    1256 non-null   float64
dtypes: float64(2)
memory usage: 29.5 KB


## Data contains some null/NaN-values.

As an easy fix, let's just drop the rows with null values. That shouldn't affect our results anyway

In [73]:
beta_data = beta_data.dropna()

In [74]:
beta_data.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 1246 entries, 2016-11-25 to 2021-11-25
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   ^OMXHCAPGI  1246 non-null   float64
 1   PON1V.HE    1246 non-null   float64
dtypes: float64(2)
memory usage: 29.2 KB


# Shift



In [75]:
import numpy as np

In [76]:
log_returns = np.log(beta_data/beta_data.shift())

# Beta calculation

In [77]:
cov = log_returns.cov()
# Variance for index
var = log_returns['^OMXHCAPGI'].var()
cov.loc['PON1V.HE', '^OMXHCAPGI']/var


0.7351913045364821

In [78]:
#!pip install -U scikit-learn

# Double checking result by calculating Beta value from Linear Regression

In [79]:
from sklearn.linear_model import LinearRegression

In [80]:
X = log_returns['^OMXHCAPGI'].iloc[1:].to_numpy().reshape(-1, 1)
Y = log_returns['PON1V.HE'].iloc[1:].to_numpy().reshape(-1, 1)
 
lin_regr = LinearRegression()
lin_regr.fit(X, Y)
 
lin_regr.coef_[0, 0]

0.735191304536483

### Looks like there's a tiny difference in the values, but not noticeable enough to warrant doing anything

* Normal beta calculation result: 0.7351913045364821
* Linear regression beta  result: 0.735191304536483

In [33]:
# pd.set_option("max_columns", None) # show all cols
# pd.set_option('max_colwidth', None) # show full width of showing cols
# pd.set_option("expand_frame_repr", False) # print cols side by side as it's supposed to be

In [31]:
hist 

Open          High           Low         Close  Volume  Dividends  Stock Splits
Date                                                                                               
2016-11-25  15104.799805  15152.419922  15064.080078  15133.219727       0          0             0
2016-11-28  15117.030273  15124.459961  14986.410156  14998.450195       0          0             0
2016-11-29  14993.980469  15021.820312  14917.059570  14965.709961       0          0             0
2016-11-30  14984.070312  15102.809570  14984.070312  15034.339844       0          0             0
2016-12-01  15031.389648  15126.400391  14983.259766  15005.419922       0          0             0
...                  ...           ...           ...           ...     ...        ...           ...
2021-11-19  29597.500000  29679.789062  29323.830078  29462.140625       0          0             0
2021-11-22  29493.259766  29615.390625  29399.130859  29521.640625       0          0             0
2021-11-23  29350.070312  29359.449219  29066.740234  29079.740234       0          0             0
2021-11-24  29049.859375  29262.609375  28918.119141  29033.259766       0          0             0
2021-11-25  29073.828125  29300.970703  29073.115234  29300.970703       0          0             0

[1247 rows x 7 columns]